verifier/helpers.go

```diff
@@ -0,0 +1,183 @@
+	"github.com/notaryproject/notation-go/verifier/truststore"
+)
+
+const (
+	// HeaderVerificationPlugin specifies the name of the verification plugin that should be used to verify the signature.
+	HeaderVerificationPlugin = "io.cncf.notary.verificationPlugin"
+
+	// HeaderVerificationPluginMinVersion specifies the minimum version of the verification plugin that should be used to verify the signature.
+	HeaderVerificationPluginMinVersion = "io.cncf.notary.verificationPluginMinVersion"
+)
+
+// VerificationPluginHeaders specifies headers of a verification plugin
+var VerificationPluginHeaders = []string{
+	HeaderVerificationPlugin,
+	HeaderVerificationPluginMinVersion,
+}
+
+var errExtendedAttributeNotExist = errors.New("extended attribute not exist")
+
+var semVerRegEx = regexp.MustCompile(`^(0|[1-9]\\d*)\\.(0|[1-9]\\d*)\\.(0|[1-9]\\d*)(?:-((?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\\.(?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\\+([0-9a-zA-Z-]+(?:\\.[0-9a-zA-Z-]+)*))?$`)
+
+func loadX509TrustStores(ctx context.Context, scheme signature.SigningScheme, policy *trustpolicy.TrustPolicy) ([]*x509.Certificate, error) {
+	var typeToLoad truststore.Type
+	switch scheme {
+	case signature.SigningSchemeX509:
+		typeToLoad = truststore.TypeCA
+	case signature.SigningSchemeX509SigningAuthority:
+		typeToLoad = truststore.TypeSigningAuthority
+	default:
+		return nil, fmt.Errorf("unrecognized signing scheme %q", scheme)
+	}
+
+	// TODO: create a package at internal/container/set for a set implementation
+	// https://github.com/notaryproject/notation-go/issues/203
+	var processedStoreSet = make(map[string]struct{})
+	var certificates []*x509.Certificate
+	x509TrustStore := truststore.NewX509TrustStore(dir.ConfigFS())
+	for _, trustStore := range policy.TrustStores {
+		if _, ok := processedStoreSet[trustStore]; ok {
+			// we loaded this trust store already
+			continue
+		}
+
+		// TODO: replace strings.Index() with Strings.Cut()
+		// https://github.com/notaryproject/notation-go/issues/204
+		i := strings.Index(trustStore, ":")
+		storeType := trustStore[:i]
+		if typeToLoad != truststore.Type(storeType) {
+			continue
+		}
+		name := trustStore[i+1:]
+		certs, err := x509TrustStore.GetCertificates(ctx, typeToLoad, name)
+		if err != nil {
+			return nil, err
+		}
+		certificates = append(certificates, certs...)
+		processedStoreSet[trustStore] = struct{}{}
+	}
+	return certificates, nil
+}
+
+// isCriticalFailure checks whether a VerificationResult fails the entire signature verification workflow.
+// signature verification workflow is considered failed if there is a VerificationResult with "Enforced" as the action but the result was unsuccessful
+func isCriticalFailure(result *notation.ValidationResult) bool {
+	return result.Action == trustpolicy.ActionEnforce && result.Error != nil
+}
+
+func verifyX509TrustedIdentitiesCore(certs []*x509.Certificate, trustPolicy *trustpolicy.TrustPolicy) error {
```

Comment:
This function `verifyX509TrustedIdentitiesCore` should be renamed to `verifyX509TrustedIdentities` and move to `verifier.go`.

The previous function `func verifyX509TrustedIdentities(trustPolicy *trustpolicy.TrustPolicy, outcome *notation.VerificationOutcome, authenticityResult *notation.ValidationResult)` should be removed.

Reply:
Then line 150
```go
verifyX509TrustedIdentities(trustPolicy, outcome, authenticityResult)
```
becomes
```go
authenticityResult.Error = verifyX509TrustedIdentities(outcome.EnvelopeContent.SignerInfo.CertificateChain, trustPolicy)
```

Reply:
> Then line 150
> 
> ```go
> verifyX509TrustedIdentities(trustPolicy, outcome, authenticityResult)
> ```
> 
> becomes
> 
> ```go
> authenticityResult.Error = verifyX509TrustedIdentities(outcome.EnvelopeContent.SignerInfo.CertificateChain, trustPolicy)
> ```

Line 150 actually needs to be:
```
err = verifyX509TrustedIdentities(outcome.EnvelopeContent.SignerInfo.CertificateChain, trustPolicy)
if err != nil {
	authenticityResult.Error = err
}
```
because authenticityResult.Error can be non-nil before line 150, and if the verification level is not enforce, the code will still reach to line 150. If line 150 returns nil error, this nil error would overwrite the previous non-nil error.